crates/ruma-client-api/src/state/send_state_event.rs

```diff
@@ -197,9 +226,28 @@ pub mod v3 {
                     (a, b, "".into())
                 };
 
+            #[cfg(feature = "unstable-msc3316")]
+            let timestamp = request
+                .uri()
+                .query()
+                .and_then(|query| {
+                    query.split('&').filter_map(|part| part.split_once('=')).find_map(
+                        |(key, value)| matches!(key, "ts" | "org.matrix.msc3316.ts").then(|| value),
+                    )
+                })
+                .map(from_json_str)
```

Comment:
JSON? That doesn't seem right. I think this should look similar to what `ruma_api!` generates, i.e. a `RequestQuery` struct that's deserialized w/ `ruma_serde::urlencoded` (the same one could be used for serialization, but might actually make the code longer).